### PR TITLE
offline mode for firewalled envs

### PR DESCRIPTION
**Pull request by stas00**
This PR implements the proposal from https://github.com/huggingface/transformers/issues/10379 to enable transformers to cache everything it needs and then run in the offline mode - e.g. in a firewalled environment.

This PR:

* [x] adds `is_offline_mode()` helper function that returns `True` when env var `TRANSFORMERS_OFFLINE` is set to `1/YES/ON`
* [x] automatically sets `local_files_only=True` in all 3 `from_pretrained()` methods
* [x] handles `ntlk` download dynamically in `run_seq2seq.py`
* [x] adds offline test (thanks to @lhoestq for the idea for mocking no network in the test) 
* [x] adds doc

This is to match the recently added `HF_DATASETS_OFFLINE=1` in `datasets` (https://github.com/huggingface/datasets/pull/1976). Tested that both work well together.

So now we can run with the network:
```
python examples/seq2seq/run_seq2seq.py --model_name_or_path t5-small --dataset_name wmt16 --dataset_config ro-en ...
```
and then with the same filesystem w/o the network or w/ a firewalled network:
```
HF_DATASETS_OFFLINE=1 TRANSFORMERS_OFFLINE=1 \
python examples/seq2seq/run_seq2seq.py --model_name_or_path t5-small --dataset_name wmt16 --dataset_config ro-en ...
```
and the latter succeeds since step 1 had all the data pre-fetched and cached.

@sgugger, @LysandreJik

### CODE REVIEW

**Comment by stas00** [comment]
So we have an issue with caching, when some files aren't there to cache, but if it didn't cache them it then fails in the offline mode.

Demo:

If you run the test with:
```
TRANSFORMERS_VERBOSITY=info pytest -sv tests/test_offline.py
```

You will see:
```
loading file https://huggingface.co/lysandre/tiny-bert-random/resolve/main/added_tokens.json from cache at None
loading file https://huggingface.co/lysandre/tiny-bert-random/resolve/main/special_tokens_map.json from cache at /home/stas/.cache/huggingface/transformers/6f94d50d8fbb665ce16eb2520f19ad218429be22b54b2f5a0af19ebe13178f83.dd8bd9bfd3664b530ea4e645105f557769387b3da9f79bdb55ed556bdd80611d
```

so it doesn't cache `added_tokens.json`

and then we run in the offlline mode it fails:
```
E                       ValueError: Connection error, and we cannot find the requested files in the cached path. Please try again or make sure your Internet connection is on.
```

How do we fix this conundrum? 

Thank you.

File: src/transformers/file_utils.py [hunk 1]
```diff
@@ -234,6 +234,13 @@
 }
 
 
+_is_offline_mode = True if os.environ.get("TRANSFORMERS_OFFLINE", "0") in ["1", "ON", "YES"] else False
```
Comment (sgugger) [inline src/transformers/file_utils.py:237]: Can we use the truthy value from the `ENV_VARS_TRUE_VALUES = {"1", "ON", "YES", "TRUE"}`? Also ignore case?

**Review by sgugger — APPROVED** [review]
Thanks for adding this!
The constant should be documented in the install page, in the section about caching models I think (https://huggingface.co/transformers/installation.html#caching-models).

**Review by LysandreJik — APPROVED** [review]
This is great, thanks for adding the tests.

**Comment by stas00** [comment]
> The constant should be documented in the install page, in the section about caching models I think (https://huggingface.co/transformers/installation.html#caching-models).

Great idea, @sgugger. Please kindly check the doc I added is good when you get a chance.

And also the kind of test I had to add is unorthodox too, so please see if it works for you. The original version couldn't have worked.

Thank you!

File: docs/source/installation.md [hunk 2]
```diff
@@ -155,6 +155,31 @@ If you expect to be downloading large volumes of models (more than 1,000) from o
 your CI setup, or a large-scale production deployment), please cache the model files on your end. It will be way
 faster, and cheaper. Feel free to contact us privately if you need any help.
 
+### Offline mode
+
+It's possible to run `transformers` in a firewalled or a no-network environment.
```
Comment (sgugger) [inline docs/source/installation.md:160]:
```suggestion
It's possible to run 🤗 Transformers in a firewalled or a no-network environment.
```

File: docs/source/installation.md [hunk 3]
```diff
@@ -155,6 +155,31 @@ If you expect to be downloading large volumes of models (more than 1,000) from o
...
+
+Setting environment variable `TRANSFORMERS_OFFLINE=1` will tell `transformers` to use local files only and will not try to look things up.
```
Comment (sgugger) [inline docs/source/installation.md:162]:
```suggestion
Setting environment variable `TRANSFORMERS_OFFLINE=1` will tell 🤗 Transformers to use local files only and will not try to look things up.
```

File: docs/source/installation.md [hunk 4]
```diff
@@ -155,6 +155,31 @@ If you expect to be downloading large volumes of models (more than 1,000) from o
...
+
+Most likely you may want to couple this with `HF_DATASETS_OFFLINE=1` that performs the same for `datasets` if you're using the latter.
```
Comment (sgugger) [inline docs/source/installation.md:164]:
```suggestion
Most likely you may want to couple this with `HF_DATASETS_OFFLINE=1` that performs the same for 🤗 Datasets if you're using the latter.
```

File: docs/source/installation.md [hunk 5]
```diff
@@ -155,6 +155,31 @@ If you expect to be downloading large volumes of models (more than 1,000) from o
...
+
+Here is an example of how this can be used on a filesystem that is shared between a normally networked and a firewalled to the external world instances.
+
+On the instance with the normal network run your program which will download and cache models (and optionally datasets if you use `datasets`). For example:
```
Comment (sgugger) [inline docs/source/installation.md:168]:
```suggestion
On the instance with the normal network run your program which will download and cache models (and optionally datasets if you use 🤗 Datasets). For example:
```

**Review by sgugger — COMMENTED** [review]
Just to be consistent with the rest of the doc.

**Pull request merged**